README.md
<p align="center"><img src="image/btcprice.png" alt="btcprice"></p>

[![Codecov](https://img.shields.io/codecov/c/github/kplachkov/btcprice.svg?style=flat-square)](https://codecov.io/gh/kplachkov/btcprice)
[![Build Status](https://img.shields.io/travis/kplachkov/btcprice.svg?style=flat-square)](https://travis-ci.org/kplachkov/btcprice)
[![Go Report Card](https://goreportcard.com/badge/github.com/kplachkov/btcprice?style=flat-square)](https://goreportcard.com/report/github.com/kplachkov/btcprice)
[![GoDoc](https://godoc.org/github.com/kplachkov/btcprice?status.svg)](https://godoc.org/github.com/kplachkov/btcprice)
[![GitHub release](https://img.shields.io/github/release/kplachkov/btcprice.svg?style=flat-square)](https://github.com/kplachkov/btcprice/releases)

**Btcprice** is a fast, simple and clean way to get the price of Bitcoin. The package uses the Blockchain API as a source of the price. Btcprice can be used with custom source of data too.

* [Installation](#installation)
* [Getting Started](#getting-started)
* [License](#license)


## Installation

### Install via `go get`

To install btcprice, use `go get`.

```bash
$ go get github.com/kplachkov/btcprice/
```

## Getting Started

```bash
$ godoc github.com/kplachkov/btcprice/
```

### Code Example:
```go
package main

import (
	"fmt"
	"log"
	"time"

	"github.com/kplachkov/btcprice"
)

// Prints the time and the price of Bitcoin.
func printBtcPriceAndTime(btcPrice float64, t time.Time) {
	fmt.Println(t.Format(time.RFC3339))
	fmt.Println("Bitcoin price:", btcPrice)
}

// Checks for changes in the price of Bitcoin.
func checkBtcPrice(timeDuration time.Duration) {
	blockchain, serErr := btcprice.NewBlockchainService()
	if serErr != nil {
		log.Fatal(serErr)
	}
	btcPrice := blockchain.BitcoinPrice.Usd.Last // The latest price of Bitcoin.
	printBtcPriceAndTime(btcPrice, time.Now())
	// Prints the time and the latest Bitcoin price if the price has changed.
	for t := range time.NewTicker(timeDuration).C {
		newBtcErr := blockchain.Update(nil)
		if newBtcErr != nil {
			log.Fatal(newBtcErr)
		}
		newBtcPrice := blockchain.BitcoinPrice.Usd.Last
		// If the price has changed, it prints the time and up to date price.
		if btcPrice != newBtcPrice {
			printBtcPriceAndTime(newBtcPrice, t)
		}
		btcPrice = newBtcPrice
	}
}

func main() {
	checkBtcPrice(time.Second) // Check the price of Bitcoin every second.
}

```

## License

BSD 3-Clause

Copyright (c) 2018-Present, kplachkov

CHANGELOG.md
# Changelog

# v0.1.1 Release - 4/8/2018

* New logo
* Minor fixes

# v0.1.0 Release - 2/4/2018

* Add support for USD
* Add the Blockchain API as a service


example_btcprice_every_minute_test.go
package btcprice_test

import (
	"fmt"
	"log"
	"time"

	"github.com/kplachkov/btcprice"
)

func Example() {
	blockchain, serErr := btcprice.NewBlockchainService()
	if serErr != nil {
		log.Fatal(serErr)
	}
	btcPrice := blockchain.BitcoinPrice.Usd.Last // The latest price of Bitcoin.
	// Prints the time and the latest Bitcoin price every minute if the price has changed.
	for t := range time.NewTicker(time.Second * 60).C {
		newBtcErr := blockchain.Update(nil) // Updates the price from the Blockchain API.
		if newBtcErr != nil {
			log.Fatal(newBtcErr)
		}
		newBtcPrice := blockchain.BitcoinPrice.Usd.Last
		// If the price has changed, it prints the time and up to date price.
		if btcPrice != newBtcPrice {
			fmt.Println(t.Format(time.RFC3339))
			fmt.Println("Bitcoin price:", newBtcPrice)
		}
		btcPrice = newBtcPrice
	}
}


src/github.com/kplachkov/btcprice/btcprice_test.go
package btcprice

import (
	"bytes"
	"io/ioutil"
	"log"
	"strconv"
	"testing"
)

// Bitcoin test price.
var BtcPriceUsdStrTest = "1.1111111"

// JSON response from the Blockchain API in bytes.
var TestJsonBytes = []byte(`{
	  "USD" : {"15m" : ` + BtcPriceUsdStrTest + `, "last" : ` + BtcPriceUsdStrTest + `, "buy" : ` + BtcPriceUsdStrTest + `, "sell" : ` + BtcPriceUsdStrTest + `, "symbol" : "$"},
	  "AUD" : {"15m" : 9289.28, "last" : 9289.28, "buy" : 9289.28, "sell" : 9289.28, "symbol" : "$"},
	  "BRL" : {"15m" : 23640.05, "last" : 23640.05, "buy" : 23640.05, "sell" : 23640.05, "symbol" : "R$"},
	  "CAD" : {"15m" : 9205.78, "last" : 9205.78, "buy" : 9205.78, "sell" : 9205.78, "symbol" : "$"},
	  "CHF" : {"15m" : 6819.59, "last" : 6819.59, "buy" : 6819.59, "sell" : 6819.59, "symbol" : "CHF"},
	  "CLP" : {"15m" : 4320355.84, "last" : 4320355.84, "buy" : 4320355.84, "sell" : 4320355.84, "symbol" : "$"},
	  "CNY" : {"15m" : 44914.1, "last" : 44914.1, "buy" : 44914.1, "sell" : 44914.1, "symbol" : "¥"},
	  "DKK" : {"15m" : 43255.64, "last" : 43255.64, "buy" : 43255.64, "sell" : 43255.64, "symbol" : "kr"},
	  "EUR" : {"15m" : 5798.39, "last" : 5798.39, "buy" : 5798.39, "sell" : 5798.39, "symbol" : "€"},
	  "GBP" : {"15m" : 5090.95, "last" : 5090.95, "buy" : 5090.95, "sell" : 5090.95, "symbol" : "£"},
	  "HKD" : {"15m" : 56151.39, "last" : 56151.39, "buy" : 56151.39, "sell" : 56151.39, "symbol" : "$"},
	  "INR" : {"15m" : 465781.43, "last" : 465781.43, "buy" : 465781.43, "sell" : 465781.43, "symbol" : "₹"},
	  "ISK" : {"15m" : 705035.71, "last" : 705035.71, "buy" : 705035.71, "sell" : 705035.71, "symbol" : "kr"},
	  "JPY" : {"15m" : 762519.36, "last" : 762519.36, "buy" : 762519.36, "sell" : 762519.36, "symbol" : "¥"},
	  "KRW" : {"15m" : 7597788.22, "last" : 7597788.22, "buy" : 7597788.22, "sell" : 7597788.22, "symbol" : "₩"},
	  "NZD" : {"15m" : 9871.73, "last" : 9871.73, "buy" : 9871.73, "sell" : 9871.73, "symbol" : "$"},
	  "PLN" : {"15m" : 24429.43, "last" : 24429.43, "buy" : 24429.43, "sell" : 24429.43, "symbol" : "zł"},
	  "RUB" : {"15m" : 410693.14, "last" : 410693.14, "buy" : 410693.14, "sell" : 410693.14, "symbol" : "RUB"},
	  "SEK" : {"15m" : 59613.93, "last" : 59613.93, "buy" : 59613.93, "sell" : 59613.93, "symbol" : "kr"},
	  "SGD" : {"15m" : 9370.27, "last" : 9370.27, "buy" : 9370.27, "sell" : 9370.27, "symbol" : "$"},
	  "THB" : {"15m" : 223157.58, "last" : 223157.58, "buy" : 223157.58, "sell" : 223157.58, "symbol" : "฿"},
	  "TWD" : {"15m" : 208329.18, "last" : 208329.18, "buy" : 208329.18, "sell" : 208329.18, "symbol" : "NT$"}
	}`)

// Test the new Blockchain structure.
func TestNewBlockchainService(t *testing.T) {
	blockchain, err := NewBlockchainService()
	if err != nil {
		t.Errorf(err.Error())
	}
	if blockchain.BitcoinPrice.Usd.Last <= 0.0 {
		t.Errorf("The latest price of Bitcoin is not set or negative.")
	}
	if blockchain.BitcoinPrice.Usd.Market <= 0.0 {
		t.Errorf("The market price of Bitcoin is not set or negative.")
	}
	if blockchain.BitcoinPrice.Usd.Sell <= 0.0 {
		t.Errorf("The sell price of Bitcoin is not set or negative.")
	}
	if blockchain.BitcoinPrice.Usd.Buy <= 0.0 {
		t.Errorf("The buy price of Bitcoin is not set or negative.")
	}
}

// Test the response of the API.
func TestGetResponse(t *testing.T) {

	blockchain, serErr := NewBlockchainService()
	if serErr != nil {
		t.Errorf(serErr.Error())
	}
	res, resErr := blockchain.getResponse()
	if resErr != nil {
		t.Errorf(resErr.Error())
	}
	statusCode := res.StatusCode
	if statusCode < 200 && statusCode > 299 {
		t.Errorf("Response status code:\ngot(%d)\nwant(%s)", statusCode, "[200 - 299]")
	}
}

// Test the fetching of the response body.
func TestFetchBytes(t *testing.T) {
	testBytes := []byte("test") // Test bytes.
	testFail := false

	blockchain, serErr := NewBlockchainService()
	if serErr != nil {
		t.Errorf(serErr.Error())
	}
	resBody := ioutil.NopCloser(bytes.NewReader(testBytes))     // The test bytes to io.ReadCloser
	resultBodyBytes, fetchErr := blockchain.fetchBytes(resBody) // The result bytes.
	if fetchErr != nil {
		t.Errorf(fetchErr.Error())
	}
	if resultBodyBytes == nil {
		testFail = true
	} else if len(resultBodyBytes) != len(testBytes) {
		testFail = true
	} else {
		// Check if the result matches the test.
		for i := range resultBodyBytes {
			if resultBodyBytes[i] != testBytes[i] {
				testFail = true
				break
			}
		}
	}
	if testFail {
		t.Errorf("Fetch response body bytes:\ngot(%s)\nwant(%s)", resultBodyBytes, testBytes)
	}
}

// Test the update of the service.
func TestUpdate(t *testing.T) {

	blockchain, serErr := NewBlockchainService()
	if serErr != nil {
		t.Errorf(serErr.Error())
	}
	// The test Bitcoin price string to float.
	btcPriceUsdFloatTest, strConvErr := strconv.ParseFloat(BtcPriceUsdStrTest, 64) // String to float64.
	if strConvErr != nil {
		log.Fatal(strConvErr)
	}

	jsonErr := blockchain.Update(TestJsonBytes)
	if jsonErr != nil {
		t.Errorf(jsonErr.Error())
	}
	// Check the latest price of Bitcoin.
	if blockchain.BitcoinPrice.Usd.Last != btcPriceUsdFloatTest {
		t.Errorf("Last price of Bitcoin:\ngot(%f)\nwant(%f)", blockchain.BitcoinPrice.Usd.Last, btcPriceUsdFloatTest)
	}
	// Check the market price of Bitcoin.
	if blockchain.BitcoinPrice.Usd.Market != btcPriceUsdFloatTest {
		t.Errorf("Market price of Bitcoin:\ngot(%f)\nwant(%f)", blockchain.BitcoinPrice.Usd.Market, btcPriceUsdFloatTest)
	}
	// Check the buy price of Bitcoin.
	if blockchain.BitcoinPrice.Usd.Buy != btcPriceUsdFloatTest {
		t.Errorf("Buy price of Bitcoin:\ngot(%f)\nwant(%f)", blockchain.BitcoinPrice.Usd.Buy, btcPriceUsdFloatTest)
	}
	// Check the sell price of Bitcoin.
	if blockchain.BitcoinPrice.Usd.Sell != btcPriceUsdFloatTest {
		t.Errorf("Sell price of Bitcoin:\ngot(%f)\nwant(%f)", blockchain.BitcoinPrice.Usd.Sell, btcPriceUsdFloatTest)
	}
}


src/github.com/kplachkov/btcprice/example_btcprice_every_minute_test.go
package btcprice_test

import (
	"fmt"
	"github.com/kplachkov/btcprice"
	"log"
	"time"
)

func Example() {
	blockchain, serErr := btcprice.NewBlockchainService()
	if serErr != nil {
		log.Fatal(serErr)
	}
	btcPrice := blockchain.BitcoinPrice.Usd.Last // The latest price of Bitcoin.
	// Prints the time and the latest Bitcoin price every minute if the price has changed.
	for t := range time.NewTicker(time.Second * 60).C {
		newBtcErr := blockchain.Update(nil) // Updates the price from the Blockchain API.
		if newBtcErr != nil {
			log.Fatal(newBtcErr)
		}
		newBtcPrice := blockchain.BitcoinPrice.Usd.Last
		// If the price has changed, it prints the time and up to date price.
		if btcPrice != newBtcPrice {
			fmt.Println(t.Format(time.RFC3339))
			fmt.Println("Bitcoin price:", newBtcPrice)
		}
		btcPrice = newBtcPrice
	}
}


src/github.com/kplachkov/btcprice/btcprice.go
// Package btcprice is a fast, simple and clean way to get the price of Bitcoin.
// The package uses the Blockchain API as a source of the price. Btcprice can be
// used with custom source of data too.
package btcprice

import (
	"encoding/json"
	"io"
	"io/ioutil"
	"net/http"
	"time"
)

// Price holds the different types of prices.
type Price struct {
	Market float64 `json:"15m"`  // 15 minutes delayed market price.
	Last   float64 `json:"last"` // The most recent market price.
	Buy    float64 `json:"buy"`  // Buy price.
	Sell   float64 `json:"sell"` // Sell price.
}

// PriceCurrency holds the price of the cryptocurrency in different fiat currencies.
type PriceCurrency struct {
	Usd Price `json:"USD"` // Prices in USD.
}

// The Blockchain API provides only the price of Bitcoin.
type Blockchain struct {
	BitcoinPrice PriceCurrency
}

// NewBlockchainService returns a new Blockchain structure with up to date prices.
func NewBlockchainService() (*Blockchain, error) {
	blockchain := Blockchain{}
	err := blockchain.Update(nil)
	return &blockchain, err
}

// Returns response from the Blockchain API.
func (Blockchain) getResponse() (*http.Response, error) {

	const url = "https://blockchain.info/ticker" // The URL of the Blockchain API.

	client := http.Client{
		Timeout: time.Second * 2, // Maximum of 2 seconds timeout.
	}

	req, reqErr := http.NewRequest(http.MethodGet, url, nil)
	if reqErr != nil {
		return nil, reqErr
	}

	req.Header.Set("btcprice", "BTC price") // Lets remote servers understand what kind of traffic they are receiving.

	res, getErr := client.Do(req)
	if getErr != nil {
		return nil, getErr
	}
	return res, nil
}

// Returns the JSON body in bytes.
func (Blockchain) fetchBytes(resBody io.ReadCloser) ([]byte, error) {

	body, readErr := ioutil.ReadAll(resBody)
	return body, readErr
}

// Update the price of Bitcoin from the Blockchain API.
// The price can be updated from a custom source too.
// Example for a normal use of Update:
// 	blockchain.BitcoinPrice.Usd.Last // 5000.4
// 	................
// 	blockchain.Update(nil) // Updates the price from the Blockchain API.
// 	blockchain.BitcoinPrice.Usd.Last // 5020.72
// Example for a custom use of Update:
//	customSource := []byte(`{
//	   "USD" : {"15m" : 7154.09, "last" : 7154.09, "buy" : 7154.09, "sell" : 7154.09, "symbol" : "$"},
//	   "AUD" : {"15m" : 9289.28, "last" : 9289.28, "buy" : 9289.28, "sell" : 9289.28, "symbol" : "$"},
//	   "BRL" : {"15m" : 23640.05, "last" : 23640.05, "buy" : 23640.05, "sell" : 23640.05, "symbol" : "R$"},
//	   .............
//	}`)
// 	blockchain.Update(customSource)
// 	blockchain.BitcoinPrice.Usd.Last // 7154.09
func (b *Blockchain) Update(manualPrice []byte) error {
	var bodyBytes []byte
	if manualPrice != nil {
		bodyBytes = manualPrice
	} else {
		var readErr error
		res, resErr := b.getResponse() // The response from the Blockchain API.
		if resErr != nil {
			return resErr
		}

		bodyBytes, readErr = b.fetchBytes(res.Body) // The JSON response body to bytes.
		if resErr != nil {
			return readErr
		}
	}
	jsonErr := json.Unmarshal(bodyBytes, &b.BitcoinPrice) // The body bytes to BitcoinPrice structure.
	return jsonErr
}
